src/Raven.Server/Documents/AI/Embeddings/EmbeddingsGenerator.cs

```diff
@@ -0,0 +1,908 @@
+
+            var generateEmbeddings = RegisterPendingEmbeddings(pending, results, value, cacheDuration);
+            return new (ReturnAsync());
+
+            async Task<ReadOnlyMemory<ReadOnlyMemory<byte>>> ReturnAsync()
+            {
+                Wake();
+                await generateEmbeddings.TaskCompletionSource.Task;
+                return new ReadOnlyMemory<ReadOnlyMemory<byte>>(generateEmbeddings.Embeddings.ToArray());
+            }
+        }
+        
+        GenerateEmbeddings RegisterPendingEmbeddings(List<string> pending, List<ReadOnlyMemory<byte>> cachedEmbeddings, string value, TimeSpan cacheDuration)
+        {
+            var newGen = new GenerateEmbeddings(_connectionStringIdentifier, cacheDuration, pending,  cachedEmbeddings,value, this);
+            if (pending.Count == 0) // all from the cache
+            {
+                newGen.TaskCompletionSource.TrySetResult();
+                return newGen;
+            }
+
+            var inCacheGen = _inFlightCache.GetOrAdd(value, newGen);
+            if (newGen == inCacheGen)
+            {
+                _work.Enqueue(newGen);
+            }
+            return inCacheGen;
+        }
+
+        public void RemoveFromCache(string value)
+        {
+            _inFlightCache.TryRemove(value, out _);   
+        }
+        
+        private bool TryGetFromCache(DocumentsOperationContext documentsContext, string text,
+            TimeSpan cacheDuration, ref List<string> expirationRefresh, out ReadOnlyMemory<byte> result)
+        {
+            var valueHash = EmbeddingsHelper.CalculateInputValueHash(text);
+            var docId = EmbeddingsHelper.GetEmbeddingCacheDocumentId(_connectionStringIdentifier, valueHash, Configuration.Quantization);
+            
+            Attachment attachment = _documentsStorage.AttachmentsStorage.GetAttachment(documentsContext, docId, valueHash, AttachmentType.Document, null);
+            if (attachment == null)
+            {
+                result = default;
+                return false;
+            }
+
+            var document = _documentsStorage.Get(documentsContext, docId);
+            if (document.Data.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata))
+            {
+                if (metadata.TryGet(Constants.Documents.Metadata.Expires, out DateTime expires))
+                {
+                    var timeToExpireSeconds = (expires - _parent._database.Time.GetUtcNow()).TotalSeconds;
+                    var halfTime = cacheDuration.TotalSeconds / 2;
+                    if (halfTime > timeToExpireSeconds)
+                    {
+                        expirationRefresh ??= [];
+                        expirationRefresh.Add(docId);
+                    }
+                }
+            }
+
+            var stream = attachment.Stream;
+            //TODO: Need to find a way to avoid this allocation in favor of pooling, etc.
```

Comment:
marking TODO

Reply:
Pretty sure we won't fix that before the release, mind